planner/core/exhaust_physical_plans.go

```diff
@@ -780,6 +780,18 @@ func (la *LogicalApply) exhaustPhysicalPlans(prop *property.PhysicalProperty) []
 	return []PhysicalPlan{apply}
 }
 
+func (p *LogicalWindowFunc) exhaustPhysicalPlans(prop *property.PhysicalProperty) []PhysicalPlan {
+	if !matchItems(prop, p.ByItems) {
+		return nil
+	}
+	window := PhysicalWindowFunc{
```

Comment:
Spark window function implementation: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExec.scala#L33
Impala window function implementation: https://github.com/cloudera/Impala/blob/cdh5-trunk/be/src/exec/analytic-eval-node.h#L34

They both require the input data to be ordered by the columns of `partition by` and `order by`, we can do this at the beginning as well. But I think it's better to `s/PhysicalWindowFunc/PhysicalStreamWindow/` to explicitly specify this required data property. And we can implement a `HashWindow` one day in the future which don't require that data property.

Reply:
In my view, it does not matter for window func whether the `partition by` is ordered. The `partition by` property is guraneetened by its' child, no matter by hash or by sort, so even if one day we support partition by hash, we just need to add one `physical hash and order operator` as the window func's child.

Reply:
The `physical hash and order operator` is the data shuffle operator? Which partitions data into different workers and sort each partition according to the specified columns?

Reply:
Basically yes, it just does what `HashWindow` will organize the input.